url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/FirstSets.java

```diff
@@ -0,0 +1,166 @@
+package fri.patterns.interpreter.parsergenerator.parsertables;
+
+import java.util.*;
+import fri.util.collections.UniqueAggregatingHashtable;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+
+/**
+	Map of all FOLLOW-sets of a SLR syntax. Key is nonterminal.
+	<p>
+	The FIRST-set for a nonterminal is the collection of terminals that
+	appear on first position of the right side of all rules where that
+	nonterminal appears on the left side.
+	<p>
+	<b>Algorithm:</b><br>
+	Search all rules that contain the nonterminal on left side.
+	Scan the right side of all these rules until a terminal or a non-nullable
+	nonterminal appears. The terminal goes to FIRST, the FIRST sets of
+	all scanned nonterminals go to FIRST. This implies that one must
+	collect FIRST sets recursively, for every scanned nonterminal.
+
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+class FirstSets extends UniqueAggregatingHashtable
+{
+	/**
+		Calculate Map of FIRST sets of all nonterminals contained in syntax.
+		The FIRST sets will then be lists within this Map that provides the
+		nonterminal as key.
+		@param syntax the grammar to scan.
+		@param nullAble Map containing Boolean true or false for nullability of every nonterminal.
+		@param nonterminals pre-built list of nonterminals
+	*/
+	public FirstSets(Syntax syntax, Nullable nullAble, List nonterminals)
+		throws ParserBuildException
+	{
+		// collect FIRST sets into this Hashtable where each
+		// nonterminal has a List of FIRST-symbols
+		Map done = new Hashtable(nonterminals.size());	// avoid recursion
+		for (int i = 0; i < nonterminals.size(); i++)	{
+			String nonterm = (String) nonterminals.get(i);
+			generateFirstSet(syntax, nullAble, nonterm, done);
+		}
+	}
+
+	private void generateFirstSet(Syntax syntax, Nullable nullAble, String nonterminal, Map done)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.